### PR TITLE
Configuration compatibility validation

### LINKED ISSUE #1238: RunAllJoined Causing Exception
If I run my benchmarks with Run or RunAll, they both work fine. If I run RunAllJoined (below) it causes an exception. I'm using version 0.11.5.1161

```
var config = DefaultConfig.Instance.With(Job.Default.With(CsProjCoreToolchain.NetCoreApp22)).With(Job.Default.With(CsProjClassicNetToolchain.Net48));
BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).RunAllJoined(config);

Unhandled Exception: System.InvalidOperationException: Sequence contains more than one element
   at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source)
   at BenchmarkDotNet.Reports.Summary.GetConfiguredOrdererOrDefaultOne(IEnumerable`1 configs)
   at BenchmarkDotNet.Reports.Summary..ctor(String title, ImmutableArray`1 reports, HostEnvironmentInfo hostEnvironmentInfo, String resultsDirectoryPath, String logFilePath, TimeSpan totalTime, ImmutableArray`1 validationErrors)
   at BenchmarkDotNet.Reports.Summary.Join(List`1 summaries, ClockSpan clockSpan)
   at BenchmarkDotNet.Running.BenchmarkRunnerClean.Run(BenchmarkRunInfo[] benchmarkRunInfos)
   at BenchmarkDotNet.Running.BenchmarkSwitcher.RunWithDirtyAssemblyResolveHelper(String[] args, IConfig config)
   at BenchmarkDotNet.Running.BenchmarkSwitcher.Run(String[] args, IConfig config)
   at BenchmarkDotNet.Running.BenchmarkSwitcher.RunAllJoined(IConfig config)
   at dotNetTips.Utility.Benchmarks.Program.Main() in C:\src\git\dotNetTips.Utility\Benchmarks\dotNetTips.Utility.Benchmarks\dotNetTips.Utility.Benchmarks\Program.cs:line 24
```

### PR DESCRIPTION
**Pull request by gsomix**
Closes #1238 

It's not complete solution, just a starting point for discussion. I've reviewed @AndreyAkinshin repro and existing validation system. It seems for now Validators work only per benchmark type and closely coupled with their configs.

I propose the following solution:

1. introduce special Validator for config compatibility checks; we need to call it from `BenchmarkRunnerClean.Validate` method in case of `JoinSummary` option enabled;
1. join all benchmark cases and Validators from all types;
1. merge all configs into one;
1. validate the cases with Validators and the config; we will get additional checks for free, e.g. `BaselineValidator`;

@adamsitnik @AndreyAkinshin What do you think? I'm looking forward to your response.

Thanks and happy #hacktoberfest!

### CODE REVIEW

**Comment by adamsitnik** [comment]
Hi @gsomix 

I like the idea of introducing a new Validator that gets all the benchmarks (`BenchmarkRunInfo[] benchmarks` from `BenchmarkRunnerClean.Validate`) and performs this special validation.

> merge all configs into one

I am not sure if this is going to be easy|possible. Merging the configs is non-trivial and you might run into handling some edge cases not related to this problem. Moving the logic that you implemented in this PR to a dedicated type should be enough. It should also make it easier for you to write the tests for it.

Thanks for working on this!

**Comment by gsomix** [comment]
Finished! CI is green (almost).

**Comment by gsomix** [comment]
@adamsitnik @AndreyAkinshin Hey, guys, any chance to get it reviewed? Thank you!

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs [hunk 1]
```diff
@@ -248,6 +248,17 @@ private static ImmutableArray<ValidationError> Validate(BenchmarkRunInfo[] bench
 
             var validationErrors = new List<ValidationError>();
 
+            if (benchmarks.Any(b => b.Config.Options.IsSet(ConfigOptions.JoinSummary)))
+            {
+                var joinedCases = benchmarks.SelectMany(b => b.BenchmarksCases).ToList();
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs]: nit: if you don't plan to add any items to the collection, an array is a better choice than list (and here `ValidationParameters` accepts an `IReadOnlyList`)

```suggestion
                var joinedCases = benchmarks.SelectMany(b => b.BenchmarksCases).ToArray();
```

File: src/BenchmarkDotNet/Validators/ConfigCompatibilityValidator.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+{
+    public class ConfigCompatibilityValidator : IValidator
+    {
+        public static readonly ConfigCompatibilityValidator FailOnError = new ConfigCompatibilityValidator();
+
+        public bool TreatsWarningsAsErrors => true;
+
+        public IEnumerable<ValidationError> Validate(ValidationParameters validationParameters)
+        {
+            var benchmarks = validationParameters.Benchmarks;
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Validators/ConfigCompatibilityValidator.cs]: nit: this local variable is used only once, it could be inlined to shorten the source code

File: src/BenchmarkDotNet/Validators/ConfigCompatibilityValidator.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+            var orderers =
+                benchmarks
+                    .Select(benchmark => benchmark.Config)
+                    .Where(config => config.Orderer != Order.DefaultOrderer.Instance)
+                    .Select(config => config.Orderer)
+                    .Distinct();
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Validators/ConfigCompatibilityValidator.cs]: nit: `.Select().Where().Select()` is a performance anti-pattern, it's better to filter first and then map which allows us to avoid an additional map operation

File: src/BenchmarkDotNet/Validators/ConfigCompatibilityValidator.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+            if (orderers.Take(2).Count() > 1)
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Validators/ConfigCompatibilityValidator.cs]: IMHO the `Take` is not needed and adds an extra complexity

```suggestion
            if (orderers.Count() > 1)
```

File: tests/BenchmarkDotNet.IntegrationTests/JoinedBenchmarksTests.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
+﻿using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Order;
+using BenchmarkDotNet.Running;
+using BenchmarkDotNet.Configs;
+using BenchmarkDotNet.Tests.Loggers;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace BenchmarkDotNet.IntegrationTests
```
Comment (adamsitnik) [inline tests/BenchmarkDotNet.IntegrationTests/JoinedBenchmarksTests.cs]: big thanks for adding the test! It looks like the benchmarks should never be executed in this case so it can be moved to the unit tests project (we try to keep the number of intergration tests low)

File: tests/BenchmarkDotNet.Tests/Validators/ConfigCompatibilityValidatorTests.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace BenchmarkDotNet.Tests.Validators
+{
+    public class ConfigCompatibilityValidatorTests
+    {
+        [Fact]
+        public void JoinedBenchmarksShouldntHaveDifferentOrderers()
```
Comment (adamsitnik) [inline tests/BenchmarkDotNet.Tests/Validators/ConfigCompatibilityValidatorTests.cs]: Could you please add one more test for a case where there are no extra orderders defined and the validator does not return any errors?

**Comment by gsomix** [comment]
@adamsitnik Thanks! I'll take a look on weekend.

**Comment by gsomix** [comment]
@adamsitnik Thanks you for the review! It's ready for the second round. :)
